### PR TITLE
[cmake] Fun with flags

### PR DESCRIPTION
**Pull request by hudokkow**
## Description
Change default -O3 flag to -O2 on Linux, OSX, BSD and RBPi and to -Os on droids and iOS for release builds.

## Motivation and Context
Convo in slack channel.

## How Has This Been Tested?
Build/run tested on linux only.

## Types of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

Original was linux only. I added the other platforms so we could discuss if it's needed/wanted or not.

https://github.com/xbmc/xbmc/commit/7a35049d09a17bc842d37067aa4603d18d495c55 is cosmetics only. It makes reading build info easier on huge logs.

ping @stefansaraev, @wsnipex and @fetzerch

### CODE REVIEW

**Comment by stefansaraev** [comment]
+1. I dont think -O3 is safe for project with big codebase (and lot of depends) like kodi.

btw, my real issue was (and still is) that cmake does not honor my $CXXFLAGS from environment (set by my build scripts). I personaly do prefer -Os on linux, I guess I can work around by appending ${CMAKE_CXX_FLAGS} to CMAKE_CXX_FLAGS_RELEASE :)

**Comment by hudokkow** [comment]
Can you upload your script? If it's top-secret, I'll be happy with only cmake calls.

**Comment by fetzerch** [comment]
Can you move it to common/ArchSetup.cmake please? I don't like that we have this code duplication.

**Comment by stefansaraev** [comment]
@hudokkow 

> Can you upload your script? If it's top-secret, I'll be happy with only cmake calls.

nothing fancy. it just exports CXXFALAGS to environment. on my env it is something like

```
-march=armv8-a -mcpu=cortex-a53 -mabi=lp64 -Wno-psabi -w -pipe -Os
```
OpenELEC and LibreELEC (and maybe other distros?) are doing the same. 

when compiling kodi, -O3 (now -O2 with this PR) was automagicaly appended at the end. rendering my -Os useless. with gcc, whatever -Ox comes last - it is used.

File: project/cmake/CMakeLists.txt [hunk 1]
```diff
@@ -460,31 +460,39 @@ configure_file(${PROJECT_SOURCE_DIR}/scripts/common/AddonHelpers.cmake
...
-  message(STATUS "#---------------------------------------------#")
-  message(STATUS "GIT_REV: ${APP_SCMID}")
-  message(STATUS "#---------------------------------------------#")
-  message(STATUS "CPACK_GENERATOR       : ${CPACK_GENERATOR}")
-  message(STATUS "CPACK_SOURCE_GENERATOR: ${CPACK_SOURCE_GENERATOR}")
-  message(STATUS "#---------------------------------------------#")
+message(STATUS "\n-- DLOPEN/DLCLOSE LIB ----------------------------------------------------------")
+message(STATUS "  ${CMAKE_DL_LIBS}")
+
+message(STATUS "\n-- FUN WITH FLAGS --------------------------------------------------------------")
```
Comment (wsnipex) [inline project/cmake/CMakeLists.txt]: COMPILER FLAGS please

File: project/cmake/CMakeLists.txt [hunk 2]
```diff
@@ -460,31 +460,39 @@ configure_file(${PROJECT_SOURCE_DIR}/scripts/common/AddonHelpers.cmake
...
+message(STATUS "  CMAKE_C_FLAGS               : ${CMAKE_C_FLAGS}")
+message(STATUS "  CMAKE_CXX_FLAGS             : ${CMAKE_CXX_FLAGS}")
+message(STATUS "  CMAKE_ASM_FLAGS             : ${CMAKE_ASM_FLAGS}")
+message(STATUS "  CMAKE_EXE_LINKER_FLAGS      : ${CMAKE_EXE_LINKER_FLAGS}\n")
+message(STATUS "  SYSTEM_LDFLAGS              : ${SYSTEM_LDFLAGS}\n")
+
+if(NOT WIN32)
+  string(TOUPPER ${CMAKE_BUILD_TYPE} BUILD_TYPE)
```
Comment (wsnipex) [inline project/cmake/CMakeLists.txt]: this var is not really needed

File: project/cmake/CMakeLists.txt [hunk 3]
```diff
@@ -460,31 +460,39 @@ configure_file(${PROJECT_SOURCE_DIR}/scripts/common/AddonHelpers.cmake
...
+  message(STATUS "  CMAKE_C_FLAGS_${BUILD_TYPE}         : ${CMAKE_C_FLAGS_${BUILD_TYPE}}")
+  message(STATUS "  CMAKE_CXX_FLAGS_${BUILD_TYPE}       : ${CMAKE_CXX_FLAGS_${BUILD_TYPE}}")
+  message(STATUS "  CMAKE_ASM_FLAGS_${BUILD_TYPE}       : ${CMAKE_ASM_FLAGS_${BUILD_TYPE}}")
+  message(STATUS "  CMAKE_EXE_LINKER_FLAGS_${BUILD_TYPE}: ${CMAKE_ASM_FLAGS_${BUILD_TYPE}}")
+
+  message(STATUS "\n-- BUILD TYPE ------------------------------------------------------------------")
+  message(STATUS "  ${BUILD_TYPE}")
```
Comment (wsnipex) [inline project/cmake/CMakeLists.txt]: please move this to the top of final message, but on one line

**Review by wsnipex — CHANGES_REQUESTED** [review]
I'm ok with setting default cflags to -O2, although we have been building with O3 for a long time without issues that boiled down to cflags.

What's not ok is to overwrite user set flags without reason, if I set my cflags I expect them to be used.

**Comment by hudokkow** [comment]
Current (master) flags we're using: http://pastebin.com/8MbAVdkM

File: project/cmake/CMakeLists.txt [hunk 4]
```diff
@@ -460,31 +460,39 @@ configure_file(${PROJECT_SOURCE_DIR}/scripts/common/AddonHelpers.cmake
...
+message(STATUS "  CMAKE_ASM_FLAGS             : ${CMAKE_ASM_FLAGS}")
+message(STATUS "  CMAKE_EXE_LINKER_FLAGS      : ${CMAKE_EXE_LINKER_FLAGS}\n")
+message(STATUS "  SYSTEM_LDFLAGS              : ${SYSTEM_LDFLAGS}\n")
+
+if(NOT WIN32)
+  string(TOUPPER ${CMAKE_BUILD_TYPE} BUILD_TYPE)
+  message(STATUS "  CMAKE_C_FLAGS_${BUILD_TYPE}         : ${CMAKE_C_FLAGS_${BUILD_TYPE}}")
+  message(STATUS "  CMAKE_CXX_FLAGS_${BUILD_TYPE}       : ${CMAKE_CXX_FLAGS_${BUILD_TYPE}}")
+  message(STATUS "  CMAKE_ASM_FLAGS_${BUILD_TYPE}       : ${CMAKE_ASM_FLAGS_${BUILD_TYPE}}")
+  message(STATUS "  CMAKE_EXE_LINKER_FLAGS_${BUILD_TYPE}: ${CMAKE_ASM_FLAGS_${BUILD_TYPE}}")
```
Comment (hudokkow) [inline project/cmake/CMakeLists.txt]: Ah? Not it doesn't. See pastebin below.

File: project/cmake/CMakeLists.txt (hunk 1 above)
Comment (hudokkow) [inline project/cmake/CMakeLists.txt]: No sense of humor... TBBT? :wink:

File: project/cmake/CMakeLists.txt (hunk 2 above)
Comment (hudokkow) [inline project/cmake/CMakeLists.txt]: Don't follow.
It's CMAKE_EXE_LINKER_FLAGS_DEBUG, not CMAKE_EXE_LINKER_FLAGS_Debug.

Last time I checked it's case sensitive.

File: project/cmake/CMakeLists.txt (hunk 3 above)
Comment (hudokkow) [inline project/cmake/CMakeLists.txt]: Okido.

**Comment by hudokkow** [comment]
@fetzerch, I'll change it to only use common/

File: project/cmake/CMakeLists.txt (hunk 4 above)
Comment (wsnipex) [inline project/cmake/CMakeLists.txt]: you're right, sorry :)

File: project/cmake/CMakeLists.txt (hunk 2 above)
Comment (wsnipex) [inline project/cmake/CMakeLists.txt]: afaik cmake should automatically use the correct ones for the build type merged with external passed flags. If it doesn't then we have to merge them ourselves.

**Comment by MartijnKaijser** [comment]
Still needed?

**Comment by hudokkow** [comment]
Nope.

**Pull request closed without merging**